Loading SQL (Structured Query Language) extension for Jupyter

In [2]:
%load_ext sql

Connecting to the database

In [3]:
%sql mysql+pymysql://root:database_password@localhost:3306/md_water_services_new

Connecting to 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

In [7]:
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

## Water Accessibility and Infrastructure 
This survey is aimed to identify the water sources people use and determine both the total and average number of users for each source in Maji Ndogo communities.
Additionally, it examined the duration citizens typically spend in queues to access water.

In [3]:
%%sql

SHOW TABLES

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

8 rows affected.

+---------------------------------+
| Tables_in_md_water_services_new |
+---------------------------------+
|         data_dictionary         |
|             employee            |
|       global_water_access       |
|             location            |
|              visits             |
|          water_quality          |
|           water_source          |
|          well_pollution         |
+---------------------------------+

# Cleaning the data

#### Cleaning well pollulation table

In [8]:
%%sql

SELECT *
FROM well_pollution
LIMIT 20

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

20 rows affected.

+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
|  source_id   |         date        |                       description                        | pollutant_ppm |  biological |         results          |
+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+
| KiRu28935224 | 2021-01-04 09:17:00 |                Bacteria: Giardia Lamblia                 |      0.0      |   495.898   | Contaminated: Biological |
| AkLu01628224 | 2021-01-04 09:53:00 |                    Bacteria: E. coli                     |      0.0      |   6.09608   | Contaminated: Biological |
| HaZa21742224 | 2021-01-04 10:37:00 |    Inorganic contaminants: Zinc, Zinc, Lead, Cadmium     |     2.715     |     0.0     |  Contaminated: Chemical  |
| HaRu19725224 | 2021-01-04 11:04:00 |                          Clean                           |   0.0288593   | 9.56996e-05 |          Clean           |
| SoRu35703224 | 2021-01-04 11:29:00 |                    Bacteria: E. coli                     |      0.0      |   22.5009   | Contaminated: Biological |
| AkHa00070224 | 2021-01-04 11:42:00 |             Inorganic contaminants: Cadmium              |    5.46739    |     0.0     |  Contaminated: Chemical  |
| HaSe21346224 | 2021-01-04 11:52:00 |                          Clean                           |   0.0140376   | 8.98989e-05 |          Clean           |
| HaYa21468224 | 2021-01-04 12:03:00 | Inorganic contaminants: Chromium, Barium, Chromium, Lead |    6.05137    |     0.0     |  Contaminated: Chemical  |
| SoRu36278224 | 2021-01-04 12:24:00 |                Parasite: Cryptosporidium                 |      0.0      |   485.162   | Contaminated: Biological |
| AkLu02155224 | 2021-01-04 12:29:00 |        Inorganic contaminants: Selenium, Arsenic         |    7.64106    |     0.0     |  Contaminated: Chemical  |
| AkRu06938224 | 2021-01-04 12:56:00 |     Inorganic contaminants: Silver, Zinc, Zinc, Lead     |    1.99096    |     0.0     |  Contaminated: Chemical  |
| AkRu05825224 | 2021-01-04 13:07:00 |              Inorganic contaminants: Silver              |    5.89926    |     0.0     |  Contaminated: Chemical  |
| AkRu05973224 | 2021-01-04 13:44:00 |                          Clean                           |    0.084344   |  6.2396e-05 |          Clean           |
| SoRu36313224 | 2021-01-04 13:46:00 |                Bacteria: Vibrio cholerae                 |      0.0      |    182.88   | Contaminated: Biological |
| KiMr24968224 | 2021-01-04 13:52:00 |                    Bacteria: E. coli                     |      0.0      |   44.2164   | Contaminated: Biological |
| KiRu27916224 | 2021-01-04 14:19:00 |               Inorganic contaminants: Zinc               |    1.06885    |     0.0     |  Contaminated: Chemical  |
| SoRu35508224 | 2021-01-04 14:49:00 |          Inorganic contaminants: Chromium, Zinc          |    4.15909    |     0.0     |  Contaminated: Chemical  |
| AkHa00706224 | 2021-01-04 16:10:00 |                          Clean                           |   0.0862577   | 4.22834e-05 |          Clean           |
| AkHa00465224 | 2021-01-04 16:28:00 |      Inorganic contaminants: Lead, Mercury, Copper       |    4.97264    |     0.0     |  Contaminated: Chemical  |
| SoRu39241224 | 2021-01-04 16:40:00 |              Inorganic contaminants: Copper              |    1.28796    |     0.0     |  Contaminated: Chemical  |
+--------------+---------------------+----------------------------------------------------------+---------------+-------------+--------------------------+

It seems like, in some cases, if the description field begins with the word “Clean”, the results have been classified as “Clean” in the results column, even though the biological column is > 0.01 
The descriptions should only have the word “Clean” if there is no biological contamination (and no chemical pol-
lutants).

In [9]:
%%sql

SELECT *
FROM well_pollution
WHERE description LIKE 'clean_%';

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

38 rows affected.

+--------------+---------------------+---------------------------------+---------------+------------+---------+
|  source_id   |         date        |           description           | pollutant_ppm | biological | results |
+--------------+---------------------+---------------------------------+---------------+------------+---------+
| AkRu06489224 | 2021-01-10 09:44:00 | Clean Bacteria: Giardia Lamblia |   0.0897904   |   38.467   |  Clean  |
| KiRu25473224 | 2021-02-07 15:51:00 | Clean Bacteria: Giardia Lamblia |   0.0630094   |  24.4536   |  Clean  |
| HaRu17401224 | 2021-03-01 13:44:00 | Clean Bacteria: Giardia Lamblia |   0.0649209   |  25.8129   |  Clean  |
| AkRu07137224 | 2021-03-04 13:41:00 | Clean Bacteria: Giardia Lamblia |   0.0656843   |  18.2978   |  Clean  |
| KiRu27205224 | 2021-03-13 14:17:00 | Clean Bacteria: Giardia Lamblia |   0.0418018   |  49.4281   |  Clean  |
| AkHa00514224 | 2021-04-11 12:11:00 | Clean Bacteria: Giardia Lamblia |   0.0305404   |  22.0255   |  Clean  |
| AmAm09776224 | 2021-05-23 11:28:00 | Clean Bacteria: Giardia Lamblia |   0.0963821   |  13.6574   |  Clean  |
| SoIl32894224 | 2021-07-11 11:37:00 | Clean Bacteria: Giardia Lamblia |   0.0712408   |  5.44957   |  Clean  |
| AkRu07366224 | 2021-07-23 11:19:00 | Clean Bacteria: Giardia Lamblia |   0.0969458   |  26.0308   |  Clean  |
| KiHa23443224 | 2021-09-05 12:34:00 | Clean Bacteria: Giardia Lamblia |     0.0828    |  13.7162   |  Clean  |
| AkRu06340224 | 2021-11-01 15:24:00 | Clean Bacteria: Giardia Lamblia |    0.051746   |  36.4594   |  Clean  |
| AkRu06363224 | 2021-11-15 13:36:00 | Clean Bacteria: Giardia Lamblia |   0.0882158   |  45.3731   |  Clean  |
| AkRu08749224 | 2021-11-19 13:37:00 |     Clean Bacteria: E. coli     |   0.0888194   |  9.28085   |  Clean  |
| AkRu07066224 | 2021-12-10 14:40:00 |     Clean Bacteria: E. coli     |   0.0151035   |  44.4652   |  Clean  |
| KiAm21996224 | 2021-12-16 10:05:00 |     Clean Bacteria: E. coli     |   0.0480415   |  25.4174   |  Clean  |
| AkRu05377224 | 2022-01-01 10:53:00 |     Clean Bacteria: E. coli     |   0.0226614   |  0.613503  |  Clean  |
| HaRu20765224 | 2022-02-07 12:39:00 |     Clean Bacteria: E. coli     |   0.0213831   |  6.95778   |  Clean  |
| SoMa34473224 | 2022-03-10 09:01:00 |     Clean Bacteria: E. coli     |   0.0745784   |  13.6458   |  Clean  |
| SoKo33474224 | 2022-03-20 16:07:00 |     Clean Bacteria: E. coli     |   0.0687177   |  10.5712   |  Clean  |
| AkHa00226224 | 2022-03-26 11:53:00 |     Clean Bacteria: E. coli     |   0.0424571   |  39.4978   |  Clean  |
| AkRu06881224 | 2022-03-28 13:37:00 |     Clean Bacteria: E. coli     |   0.0202831   |  33.0692   |  Clean  |
| AkLu02254224 | 2022-04-25 12:51:00 |     Clean Bacteria: E. coli     |   0.0635533   |  16.9482   |  Clean  |
| AkRu08416224 | 2022-05-29 15:59:00 |     Clean Bacteria: E. coli     |   0.0406701   |  27.0805   |  Clean  |
| SoMa34205224 | 2022-07-31 13:25:00 |     Clean Bacteria: E. coli     |   0.0726217   |   23.892   |  Clean  |
| KiIs23482224 | 2022-08-15 16:41:00 |     Clean Bacteria: E. coli     |   0.0539137   |  41.0842   |  Clean  |
| AmRu13252224 | 2022-08-20 14:23:00 |     Clean Bacteria: E. coli     |   0.0653589   |  30.8643   |  Clean  |
| KiRu27297224 | 2022-11-21 13:18:00 |     Clean Bacteria: E. coli     |   0.0856842   |  8.62509   |  Clean  |
| KiRu26929224 | 2022-12-18 10:52:00 |     Clean Bacteria: E. coli     |   0.0748808   |  26.6541   |  Clean  |
| HaAm16042224 | 2022-12-22 13:52:00 |     Clean Bacteria: E. coli     |   0.0605767   |  4.54755   |  Clean  |
| AkRu03107224 | 2023-01-01 15:37:00 |     Clean Bacteria: E. coli     |   0.0901101   |  18.3118   |  Clean  |
| KiHa23111224 | 2023-01-12 15:56:00 |     Clean Bacteria: E. coli     |   0.0161869   |  36.8748   |  Clean  |
| AkRu03771224 | 2023-02-26 12:40:00 |     Clean Bacteria: E. coli     |   0.0625017   |  19.7648   |  Clean  |
| AkRu04253224 | 2023-02-26 13:46:00 |     Clean Bacteria: E. coli     |   0.027

Some of the field surveyors have marked wells as Clean in the results column because the description had the word “Clean” in it, even though they have a biological contamination. So i need to find all the results that have a value greater than 0.01 in the biological column and have been set to Clean in the results column.

In [10]:
%%sql

# All records that mistakenly have 'Clean Bacteria: E.' coli are updated to Bacteria: E. coli

UPDATE
    well_pollution
SET
    description = 'Bacteria: E. coli'
WHERE
    description = 'Clean Bacteria: E. coli'

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

26 rows affected.

++
||
++
++

In [11]:
%%sql

#All records that mistakenly have Clean Bacteria: Giardia Lamblia are updated to Bacteria: Giardia Lamblia

UPDATE
  well_pollution
SET
    description = 'Bacteria: Giardia Lamblia'
WHERE
   description = 'Clean Bacteria: Giardia Lamblia'

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

12 rows affected.

++
||
++
++

In [13]:
%%sql

# Result are updated to 'Contaminated: Biological' where 'biological' is greater than 0.01 plus current results is `Clean`

UPDATE
    well_pollution
SET
   results = 'Contaminated: Biological'
WHERE
   biological > 0.01 AND results = 'Clean'

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

64 rows affected.

++
||
++
++

In [15]:
%%sql

# checking if our errors has been fixed

SELECT *
FROM
    well_pollution
WHERE
    description LIKE "Clean_%"
OR (results = "Clean" AND biological > 0.01);

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

+-----------+------+-------------+---------------+------------+---------+
| source_id | date | description | pollutant_ppm | biological | results |
+-----------+------+-------------+---------------+------------+---------+
+-----------+------+-------------+---------------+------------+---------+

We can now move on since running the query did not yeild any result. This means that the well pollution table has been cleaned.

#### Cleaning employee table

In [4]:
%%sql

Select *
from employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name  | phone_number |             email             |          address          | province_name | town_name |    position    |
+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
|          0           |  Amara Jengo   | +99637993287 |   amara.jengo@ndogowater.gov  |  36 Pwani Mchangani Road  |     Sokoto    |   Ilanga  | Field Surveyor |
|          1           |  Bello Azibo   | +99643864786 |   bello.azibo@ndogowater.gov  |   129 Ziwa La Kioo Road   |    Kilimani   |   Rural   | Field Surveyor |
|          2           |  Bakari Iniko  | +99222599041 |  bakari.iniko@ndogowater.gov  |   18 Mlima Tazama Avenue  |    Hawassa    |   Rural   | Field Surveyor |
|          3           | Malachi Mavuso | +99945849900 | malachi.mavuso@ndogowater.gov |     100 Mogadishu Road    |     Akatsi    |   Lusaka  | Field Surveyor |
|          4           |  Cheche Buhle  | +99381679640 |  cheche.buhle@ndogowater.gov  |      1 Savanna Street     |     Akatsi    |   Rural   | Field Surveyor |
|          5           | Zuriel Matembo | +99034075111 | zuriel.matembo@ndogowater.gov |  26 Bahari Ya Faraja Road |    Kilimani   |   Rural   | Field Surveyor |
|          6           |  Deka Osumare  | +99379364631 |  deka.osumare@ndogowater.gov  |    104 Kenyatta Street    |     Akatsi    |   Rural   | Field Surveyor |
|          7           | Lalitha Kaburi | +99681623240 | lalitha.kaburi@ndogowater.gov | 145 Sungura Amanpour Road |    Kilimani   |   Rural   | Field Surveyor |
|          8           |  Enitan Zuri   | +99248509202 |   enitan.zuri@ndogowater.gov  |      117 Kampala Road     |    Hawassa    |  Zanzibar | Field Surveyor |
|          10          |   Farai Nia    | +99570082739 |    farai.nia@ndogowater.gov   | 33 Angélique Kidjo Avenue |     Amanzi    |   Dahabu  | Field Surveyor |
+----------------------+----------------+--------------+-------------------------------+---------------------------+---------------+-----------+----------------+
Truncated to displaylimit of 10.

Filling the email column with the employee names and adding @ndogowater.gov. That is first_name.last_name@ndogowater.gov

In [5]:
%%sql 

## First up, i removed the space between the first and last names using REPLACE()

SELECT
    REPLACE(employee_name, ' ','.') ##Replace the space with a full stop
FROM
    employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

+---------------------------------+
| REPLACE(employee_name, ' ','.') |
+---------------------------------+
|           Amara.Jengo           |
|           Bello.Azibo           |
|           Bakari.Iniko          |
|          Malachi.Mavuso         |
|           Cheche.Buhle          |
|          Zuriel.Matembo         |
|           Deka.Osumare          |
|          Lalitha.Kaburi         |
|           Enitan.Zuri           |
|            Farai.Nia            |
+---------------------------------+
Truncated to displaylimit of 10.

In [6]:
%%sql

#Then i can used LOWER() with the result above to put it all in lower case
SELECT
    LOWER(REPLACE(employee_name, ' ','.'))
FROM
    employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

+----------------------------------------+
| LOWER(REPLACE(employee_name, ' ','.')) |
+----------------------------------------+
|              amara.jengo               |
|              bello.azibo               |
|              bakari.iniko              |
|             malachi.mavuso             |
|              cheche.buhle              |
|             zuriel.matembo             |
|              deka.osumare              |
|             lalitha.kaburi             |
|              enitan.zuri               |
|               farai.nia                |
+----------------------------------------+
Truncated to displaylimit of 10.

In [7]:
%%sql

# I then used CONCAT() to add the rest of the email address
SELECT
    CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov') AS new_email
FROM
    employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

+-------------------------------+
|           new_email           |
+-------------------------------+
|   amara.jengo@ndogowater.gov  |
|   bello.azibo@ndogowater.gov  |
|  bakari.iniko@ndogowater.gov  |
| malachi.mavuso@ndogowater.gov |
|  cheche.buhle@ndogowater.gov  |
| zuriel.matembo@ndogowater.gov |
|  deka.osumare@ndogowater.gov  |
| lalitha.kaburi@ndogowater.gov |
|   enitan.zuri@ndogowater.gov  |
|    farai.nia@ndogowater.gov   |
+-------------------------------+
Truncated to displaylimit of 10.

In [8]:
%%sql

# Here, i updated the email address in the employee table

UPDATE employee
SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')),'@ndogowater.gov')

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

++
||
++
++

In [9]:
%%sql

#Checking out the first 5 rows of the table to see if the email column has been updated
Select *
from employee
limit 5

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

5 rows affected.

+----------------------+----------------+--------------+-------------------------------+-------------------------+---------------+-----------+----------------+
| assigned_employee_id | employee_name  | phone_number |             email             |         address         | province_name | town_name |    position    |
+----------------------+----------------+--------------+-------------------------------+-------------------------+---------------+-----------+----------------+
|          0           |  Amara Jengo   | +99637993287 |   amara.jengo@ndogowater.gov  | 36 Pwani Mchangani Road |     Sokoto    |   Ilanga  | Field Surveyor |
|          1           |  Bello Azibo   | +99643864786 |   bello.azibo@ndogowater.gov  |  129 Ziwa La Kioo Road  |    Kilimani   |   Rural   | Field Surveyor |
|          2           |  Bakari Iniko  | +99222599041 |  bakari.iniko@ndogowater.gov  |  18 Mlima Tazama Avenue |    Hawassa    |   Rural   | Field Surveyor |
|          3           | Malachi Mavuso | +99945849900 | malachi.mavuso@ndogowater.gov |    100 Mogadishu Road   |     Akatsi    |   Lusaka  | Field Surveyor |
|          4           |  Cheche Buhle  | +99381679640 |  cheche.buhle@ndogowater.gov  |     1 Savanna Street    |     Akatsi    |   Rural   | Field Surveyor |
+----------------------+----------------+--------------+-------------------------------+-------------------------+---------------+-----------+----------------+

In [10]:
%%sql
# To check for the length of the phone number charater

SELECT
    LENGTH(phone_number)
FROM
    employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

+----------------------+
| LENGTH(phone_number) |
+----------------------+
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
+----------------------+
Truncated to displaylimit of 10.

The phone numbers should be 12 characters long, consisting of the plus sign, area code (99), and the phone number digits. However, when i used the LENGTH(column) function, it returns 13 characters, indicating there's an extra character.

In [11]:
%%sql

# Using the TRIM() function to remove extra chartacters in the column

SELECT
    TRIM(phone_number)
FROM
    employee;

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

+--------------------+
| TRIM(phone_number) |
+--------------------+
|    +99637993287    |
|    +99643864786    |
|    +99222599041    |
|    +99945849900    |
|    +99381679640    |
|    +99034075111    |
|    +99379364631    |
|    +99681623240    |
|    +99248509202    |
|    +99570082739    |
+--------------------+
Truncated to displaylimit of 10.

In [12]:
%%sql

# Here, i updated the correct phone number in the employee table

UPDATE employee
SET phone_number = TRIM(phone_number)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

++
||
++
++

In [13]:
%%sql
# To check for if the phone number legnth has been corected
SELECT
    LENGTH(phone_number)
FROM
    employee

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

56 rows affected.

+----------------------+
| LENGTH(phone_number) |
+----------------------+
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
|          12          |
+----------------------+
Truncated to displaylimit of 10.

Checking where employees leave

In [14]:
%%sql

SELECT 
    town_name, count(assigned_employee_id) as number_employees_in_location
FROM
    employee
GROUP BY 
     town_name

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

9 rows affected.

+-----------+------------------------------+
| town_name | number_employees_in_location |
+-----------+------------------------------+
|   Ilanga  |              3               |
|   Rural   |              29              |
|   Lusaka  |              4               |
|  Zanzibar |              4               |
|   Dahabu  |              6               |
|  Kintampo |              1               |
|   Harare  |              5               |
|  Yaounde  |              1               |
|   Serowe  |              3               |
+-----------+------------------------------+

#### Cleaning visits table

In [15]:
%%sql

SELECT *
FROM visits

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

60146 rows affected.

+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
| record_id | location_id |  source_id   |    time_of_record   | visit_count | time_in_queue | assigned_employee_id |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
|     0     |  SoIl32582  | SoIl32582224 | 2021-01-01 09:10:00 |      1      |       15      |          12          |
|     1     |  KiRu28935  | KiRu28935224 | 2021-01-01 09:17:00 |      1      |       0       |          46          |
|     2     |  HaRu19752  | HaRu19752224 | 2021-01-01 09:36:00 |      1      |       62      |          40          |
|     3     |  AkLu01628  | AkLu01628224 | 2021-01-01 09:53:00 |      1      |       0       |          1           |
|     4     |  AkRu03357  | AkRu03357224 | 2021-01-01 10:11:00 |      1      |       28      |          14          |
|     5     |  KiRu29315  | KiRu29315224 | 2021-01-01 10:17:00 |      1      |       9       |          40          |
|     6     |  AkRu05234  | AkRu05234224 | 2021-01-01 10:18:00 |      1      |       0       |          30          |
|     7     |  KiRu28520  | KiRu28520224 | 2021-01-01 10:28:00 |      1      |       0       |          34          |
|     8     |  HaZa21742  | HaZa21742224 | 2021-01-01 10:37:00 |      1      |       0       |          6           |
|     9     |  AmDa12214  | AmDa12214224 | 2021-01-01 10:58:00 |      1      |       0       |          36          |
+-----------+-------------+--------------+---------------------+-------------+---------------+----------------------+
Truncated to displaylimit of 10.

To get the most worked employee and appreciate the

In [16]:
%%sql

SELECT
    assigned_employee_id, COUNT(location_id) AS number_of_time_visited
FROM
    visits
GROUP BY 
    assigned_employee_id
ORDER BY number_of_time_visited DESC
LIMIT 3 #showing the top 3 employees that visited more locations

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

3 rows affected.

+----------------------+------------------------+
| assigned_employee_id | number_of_time_visited |
+----------------------+------------------------+
|          1           |          3708          |
|          30          |          3676          |
|          34          |          3539          |
+----------------------+------------------------+

I think these top 3 employees that visted a lot of locations should be awardsand given there accolades for a job well done.
To find out more details about them, i searched for their details in the employee table, using their assigned_employee_id

In [17]:
%%sql

SELECT 
    assigned_employee_id,employee_name, phone_number, email
FROM
    employee
WHERE
    assigned_employee_id IN(1, 30, 34)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

3 rows affected.

+----------------------+---------------+--------------+----------------------------+
| assigned_employee_id | employee_name | phone_number |           email            |
+----------------------+---------------+--------------+----------------------------+
|          1           |  Bello Azibo  | +99643864786 | bello.azibo@ndogowater.gov |
|          30          |   Pili Zola   | +99822478933 |  pili.zola@ndogowater.gov  |
|          34          |   Rudo Imani  | +99046972648 | rudo.imani@ndogowater.gov  |
+----------------------+---------------+--------------+----------------------------+

#### Analysing locations table

In [18]:
%%sql

SELECT *
FROM location

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

39650 rows affected.

+-------------+----------------------+---------------+-----------+---------------+
| location_id |       address        | province_name | town_name | location_type |
+-------------+----------------------+---------------+-----------+---------------+
|  AkHa00000  |  2 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00001  | 10 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00002  |  9 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00003  | 139 Addis Ababa Road |     Akatsi    |   Harare  |     Urban     |
|  AkHa00004  | 17 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00005  | 125 Addis Ababa Road |     Akatsi    |   Harare  |     Urban     |
|  AkHa00006  | 98 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00007  | 21 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00008  | 11 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
|  AkHa00009  |  6 Addis Ababa Road  |     Akatsi    |   Harare  |     Urban     |
+-------------+----------------------+---------------+-----------+---------------+
Truncated to displaylimit of 10.

Creating a query that counts the number of records per town

In [19]:
%%sql

SELECT
    town_name, COUNT(location_id) AS records_per_town
FROM
    location
GROUP BY town_name 

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

25 rows affected.

+-----------+------------------+
| town_name | records_per_town |
+-----------+------------------+
|   Harare  |       1650       |
|  Kintampo |       780        |
|   Lusaka  |       1070       |
|   Rural   |      23740       |
|  Abidjan  |       400        |
|   Amina   |       1090       |
|   Asmara  |       930        |
|   Bello   |       400        |
|   Dahabu  |       930        |
|   Pwani   |       520        |
+-----------+------------------+
Truncated to displaylimit of 10.

counting the records per province.

In [20]:
%%sql

SELECT
    province_name, COUNT(location_id) AS records_per_province
FROM
    location
GROUP BY province_name

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

5 rows affected.

+---------------+----------------------+
| province_name | records_per_province |
+---------------+----------------------+
|     Akatsi    |         8940         |
|     Amanzi    |         6950         |
|    Hawassa    |         6030         |
|    Kilimani   |         9510         |
|     Sokoto    |         8220         |
+---------------+----------------------+

Checking for records in towns grouped by provnce

In [21]:
%%sql

SELECT
    province_name, town_name, COUNT(location_id) AS records_per_town
FROM
    location
GROUP BY province_name, town_name
ORDER BY records_per_town DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

31 rows affected.

+---------------+-----------+------------------+
| province_name | town_name | records_per_town |
+---------------+-----------+------------------+
|     Akatsi    |   Rural   |       6290       |
|    Kilimani   |   Rural   |       5440       |
|     Sokoto    |   Rural   |       5010       |
|    Hawassa    |   Rural   |       3900       |
|     Amanzi    |   Rural   |       3100       |
|     Akatsi    |   Lusaka  |       1070       |
|    Kilimani   |   Mrembo  |       990        |
|     Amanzi    |   Asmara  |       930        |
|     Amanzi    |   Dahabu  |       930        |
|    Kilimani   |   Harare  |       850        |
+---------------+-----------+------------------+
Truncated to displaylimit of 10.

These results shows that our field surveyors did an excellent job in documenting the status of country's water crisis. Every province and town has many documented sources.
This makes me confident that the data we have is reliable enough to base our decisions on. This is an insight we can use to communicate data integrity.

Looking at the number of records for each location type

In [22]:
%%sql

SELECT location_type, COUNT(location_id) AS num_sources
FROM
    location
GROUP BY location_type
ORDER BY num_sources DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

2 rows affected.

+---------------+-------------+
| location_type | num_sources |
+---------------+-------------+
|     Rural     |    23740    |
|     Urban     |    15910    |
+---------------+-------------+

I can see that there are more rural sources than urban, but it's really hard to understand those numbers. Percentages are more relatable. Using SQL as a very overpowered calculator:
We can see that 60% of all water sources in the data set are in rural communities.

In [23]:
%%sql

SELECT ROUND(23740 / (15910 + 23740) * 100, 0) AS Per_Location  #Using the ROUND() function to round it to to a whole number


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

1 rows affected.

+--------------+
| Per_Location |
+--------------+
|      60      |
+--------------+

We can see that 60% of all water sources in the data set are in rural communities.

Insights
1. Our entire country was properly canvassed, and our dataset represents the situation on the ground.
2. 60% of our water sources are in rural communities across Maji Ndogo.

#### Diving into the source table

In [24]:
%%sql

SELECT *
FROM water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

39650 rows affected.

+--------------+----------------------+-------------------------+
|  source_id   | type_of_water_source | number_of_people_served |
+--------------+----------------------+-------------------------+
| AkHa00000224 |     tap_in_home      |           956           |
| AkHa00001224 |  tap_in_home_broken  |           930           |
| AkHa00002224 |  tap_in_home_broken  |           486           |
| AkHa00003224 |         well         |           364           |
| AkHa00004224 |  tap_in_home_broken  |           942           |
| AkHa00005224 |     tap_in_home      |           736           |
| AkHa00006224 |     tap_in_home      |           882           |
| AkHa00007224 |     tap_in_home      |           554           |
| AkHa00008224 |         well         |           398           |
| AkHa00009224 |         well         |           346           |
+--------------+----------------------+-------------------------+
Truncated to displaylimit of 10.

In [25]:
%%sql

# How many people did we survey in total?

SELECT SUM(number_of_people_served)
FROM water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

1 rows affected.

+------------------------------+
| SUM(number_of_people_served) |
+------------------------------+
|           27628140           |
+------------------------------+

In [26]:
%%sql

# How many wells, taps and rivers are there?

SELECT
    type_of_water_source, COUNT(type_of_water_source)
FROM
    water_source
GROUP BY
    type_of_water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

5 rows affected.

+----------------------+-----------------------------+
| type_of_water_source | COUNT(type_of_water_source) |
+----------------------+-----------------------------+
|     tap_in_home      |             7265            |
|  tap_in_home_broken  |             5856            |
|         well         |            17383            |
|      shared_tap      |             5767            |
|        river         |             3379            |
+----------------------+-----------------------------+

In [27]:
%%sql

# How many people share particular types of water sources on average?
SELECT
    type_of_water_source, AVG(number_of_people_served)
FROM
    water_source
GROUP BY
    type_of_water_source

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

5 rows affected.

+----------------------+------------------------------+
| type_of_water_source | AVG(number_of_people_served) |
+----------------------+------------------------------+
|     tap_in_home      |           644.0303           |
|  tap_in_home_broken  |           648.8593           |
|         well         |           278.5321           |
|      shared_tap      |          2071.3147           |
|        river         |           699.1844           |
+----------------------+------------------------------+

In [28]:
%%sql

# How many people are getting water from each type of source?
SELECT
    type_of_water_source, SUM(number_of_people_served)
FROM
    water_source
GROUP BY
    type_of_water_source


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

5 rows affected.

+----------------------+------------------------------+
| type_of_water_source | SUM(number_of_people_served) |
+----------------------+------------------------------+
|     tap_in_home      |           4678880            |
|  tap_in_home_broken  |           3799720            |
|         well         |           4841724            |
|      shared_tap      |           11945272           |
|        river         |           2362544            |
+----------------------+------------------------------+

To make it a bit simpler to interpret, decided to use percentages.
First, i used the total number of citizens, then used the result of that and divided each of the SUM(number_of_people_served) by
that number, times 100, to get percentages.

In [29]:
%%sql

SELECT
    type_of_water_source, ROUND(SUM(number_of_people_served)/27628140 * 100, 0) as percentage_people_per_source
FROM
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    percentage_people_per_source DESC

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

5 rows affected.

+----------------------+------------------------------+
| type_of_water_source | percentage_people_per_source |
+----------------------+------------------------------+
|      shared_tap      |              43              |
|         well         |              18              |
|     tap_in_home      |              17              |
|  tap_in_home_broken  |              14              |
|        river         |              9               |
+----------------------+------------------------------+

At some point, we will have to fix or improve all of the infrastructure, so we should start thinking about how we can make a data-driven decision. I think a simple approach is to fix the things that affect most people first. So using a window function, RANK() i wrote a query that ranks each type of source based on how many people in total use it.
However, If someone has a tap in their home, they already have the best source available. Since we can’t do anything more to improve this type of water source, i used a where caluse to exclude remove tap_in_home from the ranking.

In [39]:
%%sql
SELECT
    type_of_water_source, SUM(number_of_people_served) as number_of_people_served,
RANK() OVER
    (ORDER BY SUM(number_of_people_served) desc) AS Ranking
FROM
    water_source
WHERE type_of_water_source != 'tap_in_home'
GROUP BY
    type_of_water_source


Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

4 rows affected.

+----------------------+-------------------------+---------+
| type_of_water_source | number_of_people_served | Ranking |
+----------------------+-------------------------+---------+
|      shared_tap      |         11945272        |    1    |
|         well         |         4841724         |    2    |
|  tap_in_home_broken  |         3799720         |    3    |
|        river         |         2362544         |    4    |
+----------------------+-------------------------+---------+

To know exactly where to start fixing each of the water sources i used PARTITION() function to partition the water source based on the number of people served in different source_id

In [40]:
%%sql

SELECT
    source_id, type_of_water_source, number_of_people_served as number_of_people_served,
RANK() OVER
    (PARTITION BY type_of_water_source
     ORDER BY number_of_people_served desc) AS Priority_ranking
FROM
    water_source
WHERE type_of_water_source != 'tap_in_home'

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

32385 rows affected.

+--------------+----------------------+-------------------------+------------------+
|  source_id   | type_of_water_source | number_of_people_served | Priority_ranking |
+--------------+----------------------+-------------------------+------------------+
| SoRu34798224 |        river         |           998           |        1         |
| SoRu35837224 |        river         |           998           |        1         |
| SoRu36238224 |        river         |           998           |        1         |
| SoRu36791224 |        river         |           998           |        1         |
| SoRu36880224 |        river         |           998           |        1         |
| SoRu38142224 |        river         |           998           |        1         |
| SoRu37756224 |        river         |           998           |        1         |
| SoMa33775224 |        river         |           998           |        1         |
| KiRu30353224 |        river         |           998           |        1         |
| SoIl32972224 |        river         |           998           |        1         |
+--------------+----------------------+-------------------------+------------------+
Truncated to displaylimit of 10.

However, i figured out that using RANK() would not explain to the engineers where to start working because number of people served in some locations are the same. I used DENSE RANK() function to break the tile and ROW_NUMBER() to give source an identifier number

In [46]:
%%sql

SELECT
    source_id, type_of_water_source, number_of_people_served as number_of_people_served,
RANK() OVER
    (PARTITION BY type_of_water_source
     ORDER BY number_of_people_served desc) AS Priority_ranking,
DENSE_RANK() OVER
    (PARTITION BY type_of_water_source
     ORDER BY number_of_people_served desc) AS Dense_priority_ranking,
ROW_NUMBER() OVER
    (PARTITION BY type_of_water_source
     ORDER BY number_of_people_served desc) AS Row_num
FROM
    water_source
WHERE type_of_water_source != 'tap_in_home'

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

32385 rows affected.

+--------------+----------------------+-------------------------+------------------+------------------------+---------+
|  source_id   | type_of_water_source | number_of_people_served | Priority_ranking | Dense_priority_ranking | Row_num |
+--------------+----------------------+-------------------------+------------------+------------------------+---------+
| SoRu34798224 |        river         |           998           |        1         |           1            |    1    |
| SoRu35837224 |        river         |           998           |        1         |           1            |    2    |
| SoRu36238224 |        river         |           998           |        1         |           1            |    3    |
| SoRu36791224 |        river         |           998           |        1         |           1            |    4    |
| SoRu36880224 |        river         |           998           |        1         |           1            |    5    |
| SoRu38142224 |        river         |           998           |        1         |           1            |    6    |
| SoRu37756224 |        river         |           998           |        1         |           1            |    7    |
| SoMa33775224 |        river         |           998           |        1         |           1            |    8    |
| KiRu30353224 |        river         |           998           |        1         |           1            |    9    |
| SoIl32972224 |        river         |           998           |        1         |           1            |    10   |
+--------------+----------------------+-------------------------+------------------+------------------------+---------+
Truncated to displaylimit of 10.

#### I further looked into the time_in_queue column from the visits table to see how long people stay on queue to get water

In [50]:
%%sql

# To calculate how long the survey took, i used the DATEDIFF() function

SELECT DATEDIFF(MAX(time_of_record), MIN(time_of_record))
FROM visits

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

1 rows affected.

+----------------------------------------------------+
| DATEDIFF(MAX(time_of_record), MIN(time_of_record)) |
+----------------------------------------------------+
|                        924                         |
+----------------------------------------------------+

The surver took a total of 924 days, which is 2 and the half years

In [52]:
%%sql

# To see how long people queue time 
SELECT ROUND(AVG(NULLIF(time_in_queue, 0)), 0) #NULLIF would exclude the cloumns that have 0 from the calculation since tap_in_home has 0 queues
FROM visits

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

1 rows affected.

+-----------------------------------------+
| ROUND(AVG(NULLIF(time_in_queue, 0)), 0) |
+-----------------------------------------+
|                   123                   |
+-----------------------------------------+

On a average, we have 123 minutes which is about 2 hours This may sound reasonable, but some days might have more people who need water, and only have time to go and collect some on certain days.

In [55]:
%%sql

#To see the queue times aggregated across the different days of the week.

SELECT DAYNAME(time_of_record) as Day_of_week, ROUND(AVG(NULLIF(time_in_queue, 0)), 0) as avg_queue_time
FROM visits
GROUP BY DAYNAME(time_of_record)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

7 rows affected.

+-------------+----------------+
| Day_of_week | avg_queue_time |
+-------------+----------------+
|    Friday   |      120       |
|   Saturday  |      246       |
|    Sunday   |       82       |
|    Monday   |      137       |
|   Tuesday   |      108       |
|  Wednesday  |       97       |
|   Thursday  |      105       |
+-------------+----------------+

Saturdays have much longer queue times compared to the other days

In [58]:
%%sql

#To see the queue times aggregated across the different hour in a day.

SELECT Hour(time_of_record) as hour_of_day, ROUND(AVG(NULLIF(time_in_queue, 0)), 0) as avg_queue_time
FROM visits
GROUP BY Hour(time_of_record)

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

14 rows affected.

+-------------+----------------+
| hour_of_day | avg_queue_time |
+-------------+----------------+
|      9      |      118       |
|      10     |      114       |
|      11     |      111       |
|      12     |      112       |
|      13     |      115       |
|      14     |      114       |
|      15     |      114       |
|      16     |      114       |
|      7      |      149       |
|      18     |      147       |
+-------------+----------------+
Truncated to displaylimit of 10.

The hour number is difficult to interpret. A format like 06:00 will be easier to read, so i changed the format

In [59]:
%%sql

#To see the queue times aggregated across the different hour in a day.

SELECT TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day, ROUND(AVG(NULLIF(time_in_queue, 0)), 0) as avg_queue_time
FROM visits
GROUP BY hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

14 rows affected.

+-------------+----------------+
| hour_of_day | avg_queue_time |
+-------------+----------------+
|    09:00    |      118       |
|    10:00    |      114       |
|    11:00    |      111       |
|    12:00    |      112       |
|    13:00    |      115       |
|    14:00    |      114       |
|    15:00    |      114       |
|    16:00    |      114       |
|    07:00    |      149       |
|    18:00    |      147       |
+-------------+----------------+
Truncated to displaylimit of 10.

To properly understand and generate insigts from this information, i decided to write a query use a CASE() function to breakdown the queue time for each hour of each day of the week. For rows, i used the hour of the day in the previous time format query, and then make each column a different day.

In [61]:
%%sql

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
# Sunday
ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
ELSE NULL
    END),0) AS Sunday,

# Monday
ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue
ELSE NULL
    END),0) AS Monday,

# Tuesday
ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue
ELSE NULL
    END),0) AS Tuesday,

# Wednesday
ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue
ELSE NULL
    END),0) AS Wednesday,

# Thursday
ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue
ELSE NULL
    END),0) AS Thursday,

# Friday
ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue
ELSE NULL
    END),0) AS Friday,

# Saturday
ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue
ELSE NULL
    END),0) AS Saturday

FROM
    visits
WHERE
    time_in_queue != 0 # this excludes other sources with 0 queue times
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day

Running query in 'mysql+pymysql://root:***@localhost:3306/md_water_services_new'

14 rows affected.

+-------------+--------+--------+---------+-----------+----------+--------+----------+
| hour_of_day | Sunday | Monday | Tuesday | Wednesday | Thursday | Friday | Saturday |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
|    06:00    |   79   |  190   |   134   |    112    |   134    |  153   |   247    |
|    07:00    |   82   |  186   |   128   |    111    |   139    |  156   |   247    |
|    08:00    |   86   |  183   |   130   |    119    |   129    |  153   |   247    |
|    09:00    |   84   |  127   |   105   |     94    |    99    |  107   |   252    |
|    10:00    |   83   |  119   |    99   |     89    |    95    |  112   |   259    |
|    11:00    |   78   |  115   |   102   |     86    |    99    |  104   |   236    |
|    12:00    |   78   |  115   |    97   |     88    |    96    |  109   |   239    |
|    13:00    |   81   |  122   |    97   |     98    |   101    |  115   |   242    |
|    14:00    |   83   |  127   |   104   |     92    |    96    |  110   |   244    |
|    15:00    |   83   |  126   |   104   |     88    |    92    |  110   |   248    |
+-------------+--------+--------+---------+-----------+----------+--------+----------+
Truncated to displaylimit of 10.

1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesday has the lowest queue times, but long queues on Wednesday evening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing for water, perhaps for the week's supply?
4. The shortest queues are on Sundays, and this is a cultural thing. Proparbly because most people prioritise family and religion, so Sundays are spent with family and friends.

## Insights

1. Most water sources are rural (60%).
2. 43% of our people are using shared taps.
3. 31% of our population has water infrastructure in their homes, but within that group, 45% face non-functional systems due to issues with pipes, pumps, and reservoirs.
4. 18% of our people are using wells.
5. Our citizens often face long wait times for water, averaging more than 120 minutes.
6. In terms of queues:
- Queues are very long on Saturdays.
- Queues are longer in the mornings and evenings.
- Wednesdays and Sundays have the shortest queues.
7. Most water sources are in rural areas. We need to ensure our teams know this as this means they will have to make these repairs/upgrades in rural areas where road conditions, supplies, and labour are harder challenges to overcome.

## Recommendations

We want to focus our efforts on improving the water sources that affect the most people.
- Most people will benefit if we improve the shared taps first.
- Fixing existing infrastructure will help many people. If they have running water again, they won't have to queue, thereby shorting queue times for others. So we can solve two problems at once.
- Installing taps in homes will stretch our resources too thin, so for now, if the queue times are low, we don't have to improve that source.